libgalois/include/katana/analytics/ClusteringImplementationBase.h

```diff
@@ -0,0 +1,680 @@
+/*
+ * This file belongs to the Galois project, a C++ library for exploiting
+ * parallelism. The code is being released under the terms of the 3-Clause BSD
+ * License (a copy is located in LICENSE.txt at the top-level directory).
+ *
+ * Copyright (C) 2018, The University of Texas at Austin. All rights reserved.
+ * UNIVERSITY EXPRESSLY DISCLAIMS ANY AND ALL WARRANTIES CONCERNING THIS
+ * SOFTWARE AND DOCUMENTATION, INCLUDING ANY WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR ANY PARTICULAR PURPOSE, NON-INFRINGEMENT AND WARRANTIES OF
+ * PERFORMANCE, AND ANY WARRANTY THAT MIGHT OTHERWISE ARISE FROM COURSE OF
+ * DEALING OR USAGE OF TRADE.  NO WARRANTY IS EITHER EXPRESS OR IMPLIED WITH
+ * RESPECT TO THE USE OF THE SOFTWARE OR DOCUMENTATION. Under no circumstances
+ * shall University be liable for incidental, special, indirect, direct or
+ * consequential damages or loss of profits, interruption of business, or
+ * related expenses which may arise from use of Software or Documentation,
+ * including but not limited to those resulting from defects in Software and/or
+ * Documentation, or loss or inaccuracy of data of any kind.
+ */
+
+#ifndef KATANA_LIBGALOIS_KATANA_ANALYTICS_CLUSTERINGIMPLEMENTATIONBASE_H_
+#define KATANA_LIBGALOIS_KATANA_ANALYTICS_CLUSTERINGIMPLEMENTATIONBASE_H_
+
+#include <fstream>
+#include <iostream>
+#include <random>
+
+#include "katana/AtomicHelpers.h"
+#include "katana/Galois.h"
+#include "katana/LargeArray.h"
+#include "katana/analytics/Utils.h"
+
+namespace katana::analytics {
+
+// Maintain community information
+template <typename EdgeWeightType>
+struct CommunityType {
+  std::atomic<uint64_t> size;
+  std::atomic<EdgeWeightType> degree_wt;
+  EdgeWeightType internal_edge_wt;
+};
+
+typedef katana::LargeArray<uint64_t> LargeArray;
+
+using LCPreviousCommunityId = katana::PODProperty<uint64_t>;
+using LCCurrentCommunityId = katana::PODProperty<uint64_t>;
+using LCColorId = katana::PODProperty<int64_t>;
+template <typename EdgeWeightType>
+using LCDegreeWeight = katana::PODProperty<EdgeWeightType>;
+
+template <typename EdgeWeightType>
+using LCEdgeWeight = katana::PODProperty<EdgeWeightType>;
+
+template <typename _Graph, typename _EdgeType, typename _CommunityType>
+struct ClusteringImplementationBase {
+  using Graph = _Graph;
+  using GNode = typename Graph::Node;
+  using EdgeTy = _EdgeType;
+  using CommunityType = _CommunityType;
+
+  /*
+ * Typedefs
+ */
+  constexpr static const uint64_t INF_VAL =
+      std::numeric_limits<uint64_t>::max() / 2 - 1;
+  constexpr static const uint64_t UNASSIGNED =
+      std::numeric_limits<uint64_t>::max();
+  constexpr static const double DOUBLE_MAX =
+      std::numeric_limits<double>::max() / 4;
+
+  typedef katana::LargeArray<uint64_t> LargeArray;
+  typedef katana::LargeArray<EdgeTy> largeArrayEdgeTy;
```

Comment:
These should have better names or just be removed. The original type isn't that long. You can `using katana::LargeArray` to shorten the types a bit without hiding the element types.

Also `largeArrayEdgeTy` should be capitalized.